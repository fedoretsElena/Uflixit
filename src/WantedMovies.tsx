import React, { Component } from 'react';
import axios from 'axios';

import ErrorMsg from './shared/ErrorMsg';
import Title from './shared/Title';
import Loader from './shared/Loader';
import MediaList from './shared/MediaList';
import ApiConfig from './core/ApiConfig';

interface IState {
    movies: any[],
    loaded: boolean,
    loading: boolean
    error: string | null,
}

class WantedMovies extends Component {
    state: IState = {
        movies: [],
        error: null,
        loaded: false,
        loading: false
    };

    render() {
        const {loaded, loading, error, movies} = this.state;

        return (
            <>
                <Title title='Wanted movies'
                       length={movies.length} />
                {loading && ( <Loader/> )}
                {loaded && movies && ( <MediaList items={movies}
                                                  type={'movies'} />)}
                {error && !loading && (<ErrorMsg msg={error} />)}
            </>
        );
    }

    componentDidMount(): void {

        this.getWantedMoviesIds()
            .then(res => this.getWantedMoviesPosters(res));
    }

    private getWantedMoviesIds(): Promise<any> {
        this.setState({
            loading: true
        });

        return axios.get(ApiConfig.getWantedMoviesPath);
    }

    private getWantedMoviesPosters(ids: string[]): void {
        let movies: { id?: string, poster?: string }[] = [...ids.map(id => {
            return {id}
        })];

        axios.all(ids.map(id => axios.get(ApiConfig.getMoviePosterPath.replace('{id}', id))))
            .then((res: any) => {

                res.forEach((item: string, index: number) => {
                    movies[index].poster = item;
                });

                this.setState({
                    movies,
                    error: null,
                    loaded: true,
                    loading: false
                });
            })
            .catch(() => this.setState({
                    loading: false,
                    error: 'Can not get posters of movies using ids.'
                })
            );
    }
}

export default WantedMovies;
