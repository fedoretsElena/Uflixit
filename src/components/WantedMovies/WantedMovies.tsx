import React, { Component } from "react";

import BaseMedia from "../../models/BaseMedia";
import MediaService from "../../data/services/mediaService";
import ErrorMsg from "../shared/ErrorMsg";
import Loader from "../shared/Loader";
import MediaList from "../shared/MediaList";
import Title from "../shared/Title";

interface IState {
    movies: BaseMedia[];
    loaded: boolean;
    loading: boolean;
    error: string | null;
}

class WantedMovies extends Component {
    state: IState = {
        movies: [],
        error: null,
        loaded: false,
        loading: false
    };
    mediaService: any = new MediaService();

    render() {
        const { loaded, loading, error, movies } = this.state;

        return (
            <>
                <Title title="Wanted movies" length={movies.length} />
                {loading && <Loader />}
                {loaded && movies && <MediaList items={movies} />}
                {error && !loading && <ErrorMsg msg={error} />}
            </>
        );
    }

    componentDidMount(): void {
        this.getWantedMovies();
    }

    private getWantedMovies(): void {
        this.setState({
            loading: true
        });

        this.mediaService
            .getWantedMoviesIds()
            .then((movies: BaseMedia[]) => {
                this.setState({
                    movies,
                    error: null,
                    loaded: true,
                    loading: false
                });
            })
            .catch(() =>
                this.setState({
                    loading: false,
                    error: "Can not get posters of movies using ids."
                })
            );
    }
}

export default WantedMovies;
