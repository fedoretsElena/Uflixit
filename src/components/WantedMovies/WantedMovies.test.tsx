import WantedMovies from "./WantedMovies";

jest.mock("../../data/services/mediaService");
jest.mock("../shared/MediaList");

describe("WantedMovies component", () => {
    it("render correctly WantedMovies component", () => {
        const TextInputComponent = shallow(<WantedMovies />);

        expect(TextInputComponent).toMatchSnapshot();
    });

    it("check is preloader working", () => {
        const component = mount(<WantedMovies />);
        const loader = component.find(".whirlpool");

        expect(loader).toHaveLength(1);
    });

    it("should fetch data from mock service and mount them", done => {
        const component = shallow(<WantedMovies />);

        setTimeout(() => {
            component.update();

            expect(component.instance().state.movies.length).toEqual(2);
            expect(component.instance().state.movies[1].title).toEqual(
                "The walking Dead"
            );
            done();
        });
    });
});
