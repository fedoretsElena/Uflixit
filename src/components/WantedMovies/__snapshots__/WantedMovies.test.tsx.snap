// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WantedMovies component render correctly WantedMovies component 1`] = `
<Fragment>
  <Title
    length={0}
    title="Wanted movies"
  />
  <Loader />
</Fragment>
`;
